### PR TITLE
fix(dynamodb): autoscale dynamo global secondary indexes

### PR DESCRIPTION
**Pull request by jpetto**
# Goal

add autoscaling to dynamodb global secondary indexes.

## Reference

Tickets:
* https://getpocket.atlassian.net/browse/BACK-499

## Implementation Decisions

please check the `TODO`s in the code to validate my defaults/assumptions.

### CODE REVIEW

File: src/base/ApplicationDynamoDBTable.ts
Comment (kkyeboah) [inline src/base/ApplicationDynamoDBTable.ts:155]: I think it should be the default behavior but not forced.

File: src/base/ApplicationDynamoDBTable.ts
Comment (kkyeboah) [inline src/base/ApplicationDynamoDBTable.ts:150]: My vote is yes

File: src/base/ApplicationDynamoDBTable.ts
Comment (kkyeboah) [inline src/base/ApplicationDynamoDBTable.ts:161]: Looks like line 161 here to 193 could be extracted to a method and reused for both table and index autoscaling

**Review by kkyeboah — CHANGES_REQUESTED** [review]
Looks good to me. I just have a refactoring comment.

File: src/base/ApplicationDynamoDBTable.ts
Comment (jpetto) [reply src/base/ApplicationDynamoDBTable.ts:155]: so the question becomes - do we extend the `DynamodbTableGlobalSecondaryIndex` interface to include a `maxCapacity` property? i'm not a huge fan of extending third-party indexes, but this does seem to be a gap in functionality/configurability.

(note that the `DynamodbTableGlobalSecondaryIndex` interface defines a `writeCapacity` and `readCapacity` - which i think makes sense to use as the `minCapacity` - but happy to hear other thoughts if we need to extend this interface with both `minCapacity` and `maxCapacity`.)

**Comment by jpetto** [comment]
has been validated with `terraform apply` against our dev infra as of right now.

File: src/base/ApplicationDynamoDBTable.ts
Comment (bassrock) [inline src/base/ApplicationDynamoDBTable.ts:60]: @jpetto that's concerning given this [warning on dynamodb in terraform ](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/dynamodb_table )

Perhaps a cdk bug?

File: src/base/ApplicationDynamoDBTable.ts
Comment (bassrock) [inline src/base/ApplicationDynamoDBTable.ts:207]: Depends on like lifecycle above are baseline terraform properties. Perhaps the cdk renamed them?

**Review by bassrock — CHANGES_REQUESTED** [review]
I have some concerns about lifecycle and depends on disappearing. Can we see if these are bugs in the cdk or if we can escape hatch them?

**Review by chelseatroy — COMMENTED** [review]
Happy to review but this PR seems a few weeks old so wanted to make sure we're still wanting to merge this?

**Comment by jpetto** [comment]
> Happy to review but this PR seems a few weeks old so wanted to make sure we're still wanting to merge this?

this has been bumped in terms of priority. hold off on reviewing for now.

**Review by bassrock — APPROVED** [review]
approving to unblock my previous review. but @Pocket/backend someone else should approve.

**Review by mmiermans — APPROVED** [review]
lgtm :raised_hands:

**Review by kkyeboah — APPROVED** [review]
👍

**Pull request merged**
